### DISCUSSION TITLE
G-API Alumni Forum

### DISCUSSION THREAD

**Discussion by dmatveev** [g-api-org/discussions#1] [General]
Moving our bi-weekly notes to Github Discussions

Old notes are available [here](https://github.com/orgs/g-api-org/teams/core-team/members/archived_team_posts.json).

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2023 May 23

## Meeting minutes

### Attendees

- @dmatveev 
- Ruslan
- @TolyaTalamanov (synced offline)

### Agenda

#### OpenCV 4.8 Release

@dmatveev  - selected PRs to merge:

- Anna’s old PRs: [#21685](https://github.com/opencv/opencv/pull/21685),  [#21797](https://github.com/opencv/opencv/pull/21797)
- ONNX & Python-related changes: [#23597](https://github.com/opencv/opencv/pull/23597), [#20579](https://github.com/opencv/opencv/pull/20579)
- OpenVINO 2.0 backend (partial): [#23595](https://github.com/opencv/opencv/pull/23595) 
- <Possible fix in IE/OV1.0 backend - Anatoliy TBD>
- <Fixes to existing reported problems (MKL, etc) - Dmitry TBD>
- Optional (if time permits to complete): Multi-stream [#23334](https://github.com/opencv/opencv/pull/23334)

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2023 Jun 06

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

1. @dmatveev - INTC
2. @TolyaTalamanov - INTC

### Agenda

- Discussed OpenCV 4.8 release status, all we wanted is merged
- Discussed OpenVINO 2.0 backend status. The foundational changes are merged. What's remaining: infer-list, infer-2. infer-roi may not be required.

#### OpenCV 4.8 Release

* Anna’s old PRs: https://github.com/opencv/opencv/pull/21685, https://github.com/opencv/opencv/pull/21797
* ONNX & Python-related changes: https://github.com/opencv/opencv/pull/23597, https://github.com/opencv/opencv/pull/20579
* OpenVINO 2.0 backend (partial): https://github.com/opencv/opencv/pull/23595
* <Possible fix in IE/OV1.0 backend - Anatoliy TBD>
* <Fixes to existing reported problems (MKL, etc) - Dmitry TBD>
* Optional (if time permits to complete): Multi-stream https://github.com/opencv/opencv/pull/23334

Last-minute stuff to add:
- infer-list in OV2.0 backend
- fix for OV1.0 backend w.r.t. input images (continuation of https://github.com/opencv/opencv/pull/23668)

#### GSoC - 2023

* @TolyaTalamanov - discussing the roadmap with the contributor.
* @dmatveev - meeting with contributor later today, will check for updates. So far repo is created at https://github.com/g-api-org/tutorials-py

#### Personal updates

### Future agenda

* GSoC
* Plugins for heavy dependencies
* ONNX RT with OV UEP and Tensor RT - device (EP) selection.
* DNN backend - e.g. https://github.com/opencv/opencv_zoo/tree/master/models/handpose_estimation_mediapipe
* Tracking[?]

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2023 Aug 15

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* Dmitry - INTC
* Anatoliy - INTC
* Ruslan - HWEI

### Agenda

* Updated 5.0 roadmap (added more items)
* Discussed moving preprocessing out of inference backends. May be not that trivial now. A lot of corner cases. Probably a new OCV DNN backend could be written "as is" now to avoid blocking with the implementation
* TBB executor - complex and never complete, may be replaced with something smaller but what actually works.

#### Personal updates

n/a

### Future agenda

Dmitry to turn 5.0 roadmap into Discussions.

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2023 Oct 10

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* Dmitry - INTC

### Agenda

#### Personal updates

*Dmitry*

- Fixing ADE build (once again): https://github.com/opencv/ade/pull/40 + related PRs
- Introduced CV over OV (early stage PoC but looks promising): https://github.com/opencv/opencv/pull/24379
- Updated 5.0 roadmap sections
  - To fix: https://github.com/opencv/opencv/wiki/G-API-Ideas-towards-OpenCV-5.0#what-wed-like-to-fix
  - To remove: https://github.com/opencv/opencv/wiki/G-API-Ideas-towards-OpenCV-5.0#what-wed-like-to-remove

### Future agenda

TBC

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2023 Nov 7

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @dmatveev , @TolyaTalamanov 

### Agenda

#### PR Status for OpenCV 4.9

##### Confident PRs:

- https://github.com/opencv/opencv/pull/24224

  - WIP, PR is red (needs test update)

- https://github.com/opencv/opencv/pull/24060 

  - Stuck mainly with CMake-level changes.

##### RIsky PRs:

- https://github.com/opencv/opencv/pull/24068

  - Changes are done but unable to test. No suitable machine available. ⚠️Merge conflicts in place

- https://github.com/opencv/opencv/pull/23334

  - API + tests only, implementation isn't there yet. Dmitry will implement as best effort during 4.9 timeframe (hopefully)

- https://github.com/opencv/opencv/pull/20579

  - Stuck in merge 2 years ago w/o any major issues, needs to be cleaned up and merged. Added into 4.9

##### PRs that needed but don't exist

- https://github.com/opencv/opencv/issues/22736
- https://github.com/opencv/opencv/issues/23578

#### On plugins

- Approach via C++ / dlopen

  - pyopencv.so is loaded in Python, G-API makes lookup for libraries like OV, ONNX, VPL, etc... to dispatch calls into it
  - Individual G-API components try to load external dependencies dynamically and fail if they couldn't. The API contract is a question here.

- Approach via Python

  - Implement new backends via Python. G-API won't know about the external dependencies, they will come from the Python side as callbacks
  - Pros: any new API can be externally linked into G-API
  - Cons: performance + need to expose the API internals to allow implementing backends. Graph structure, to name a few.

### Future agenda

Continue w/4.9 & 5.0 plans.

**Comment by TolyaTalamanov** [comment]
# G-API Alumni Forum -- 2023 Nov 21

## Meeting protocol

## Meeting minutes

### Attendees

* @rgarnov, @TolyaTalamanov 

### Agenda

#### PR Status for OpenCV 4.9

- https://github.com/opencv/opencv/pull/24060 

  - Merged

- https://github.com/opencv/opencv/pull/24068

  - @rgarnov volunteered to validate changes on his machine. 
  - AR: @TolyaTalamanov resolve merge conflicts and provide instructions to @rgarnov

- https://github.com/opencv/opencv/pull/20579

  - No updates so far....

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2023 Dec 5

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @dmatveev, @TolyaTalamanov 

### Agenda

#### Personal updates

* @TolyaTalamanov - trying to bump ONNX RT version to 0.16, but it requires C++14. OpenCV forces C++11: https://github.com/opencv/opencv/pull/24601
  * Check if `ENABLE_CXX14` (or `ENABLE_C++14`) option exist at CMake level.

* @smirnov-alexey fixed several issues in the IE 1.0 backend to fix OMZ demos, but we need to update our demos to the OV backend (2.0):
  * https://github.com/opencv/opencv/pull/24615
  * Open: when do we deprecate the IE 1.0 API?

* Vadim contributed 0d/1d data types for DNN so several G-API tests were failed and so disabled: https://github.com/opencv/opencv/pull/23473
  * @TolyaTalamanov to come up with test cases which we've worked on - and check if it affects our tooling.

New bugs:

* https://github.com/opencv/opencv/issues/24431
  * @TolyaTalamanov - to follow-up 

### Future agenda

**Comment by TolyaTalamanov** [comment]
# G-API Alumni Forum -- 2023 Dec 19

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @rgarnov, @TolyaTalamanov 

### Agenda

#### Personal updates

* @TolyaTalamanov - [ONNXRT version bump](https://github.com/opencv/opencv/pull/24601) is merged
### Future agenda

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2024 Jan 16

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @dmatveev , @TolyaTalamanov 
* @rgarnov

### Agenda

#### OpenCV 4.9 is out!

* https://github.com/opencv/opencv/wiki/ChangeLog#version490
* Includes OT - closes one of our 5.0 plans! 🎉🎉🎉

#### Personal updates

* @TolyaTalamanov
  - A new executor https://github.com/opencv/opencv/pull/24845

* @rgarnov
  - Mentioned https://github.com/dmlc/dlpack as a new DL data structure for all frameworks

### Future agenda

* @TolyaTalamanov - share drafts of infer manifesto here: https://github.com/orgs/g-api-org/discussions/14
* @TolyaTalamanov - align on the threaded executor implementation https://github.com/orgs/g-api-org/discussions/13

**Comment by TolyaTalamanov** [comment]
# G-API Alumni Forum -- 2024 Jan 30

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @dmatveev , @TolyaTalamanov 

### Agenda

#### Personal updates

* @TolyaTalamanov
  - A new executor https://github.com/opencv/opencv/pull/24845

### Future agenda

* @TolyaTalamanov - share drafts of infer manifesto here: https://github.com/orgs/g-api-org/discussions/14
* @TolyaTalamanov - align on the threaded executor implementation https://github.com/orgs/g-api-org/discussions/13
* @TolyaTalamanov - Revive python bindings for `QueueSource`

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2024 Feb 13

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @dmatveev , @TolyaTalamanov 

### Agenda

* GSoC 2024: https://github.com/opencv/opencv/wiki/GSoC_2024 - call for project ideas

* OpenCV 5.0 development started
  - https://github.com/orgs/opencv/projects/1/views/1
  - Some breaking changes, Vadim's PR  https://github.com/opencv/opencv/pull/24962


#### Personal updates

* @dmatveev  - High-order constructs discussion is started:
  - https://github.com/orgs/g-api-org/discussions/16
  - Discussed past (previously unpublished) MTCNN experience: high-order constructs like `collect<>` and `map<>` were introduced there, and run nested `GComputation`s as part of the larger graph (were opaque to the parent graph's executors, just like user kernels).

### Future agenda

* @TolyaTalamanov 
  - understand why Python's `GInferInputs` is not a dict.
  - propose how to update/pass-around `GInferInputs` in transformer-like scenarios.

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2024 Mar 12

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @dmatveev, @TolyaTalamanov 

### Agenda

#### Personal updates

- @dmatveev - Merged a couple of PRs for 5.0
  - Makings tests lighter: https://github.com/opencv/opencv/pull/25060
  - Value-initialized GMat https://github.com/opencv/opencv/pull/25055 - a base prerequisite for:
    - In-graph weights (for DNN-like engines)
    - Accumulated execution (T-API-like, but for expressions)

- @TolyaTalamanov - no updates so far
  - https://github.com/opencv/opencv/pull/25054 -  Fixed test hang by lowering the IE version
    - Alexander reports hangs again - like here: https://github.com/opencv/opencv/actions/runs/8153782389/job/22285849425?pr=25149
    - Likely it is a VideoCapture problem, not G-API's - as investigated earlier by @dmatveev 

- Required updates:
  - https://github.com/opencv/opencv/pull/22469 - Resize in AVX - opened by @anna-khakimova back in 2022
  - https://github.com/opencv/opencv/pull/20579 - `compile()` in Python - Got conflicts, need rebase(?)

### Future agenda

- 5.0 plans continued?
- DNN module for infer - @dmatveev - WIP
- Preprocessing migration - @dmatveev - WIP
- HAL prototype - @dmatveev - WIP

**Comment by dmatveev** [comment]
# G-API Alumni Forum -- 2024 Apr 09

## Meeting protocol

- Participation is voluntary.
- Respect your company's NDA! Don't share any roadmaps, plans, and IP.
- Discuss only OpenCV and G-API **public** efforts and questions.

## Meeting minutes

### Attendees

* @dmatveev, @TolyaTalamanov 
* @rgarnov 

### Agenda

#### Incoming issue reports from OpenCV

* https://github.com/opencv/opencv/issues/25053 - hang in streaming-related tests. Need to prepare a VideoCapture-based reproducer (issue was reopened due to recent occurences)
* https://github.com/opencv/opencv/pull/25321 - related but now drafted (need follow-up from us)
* https://github.com/opencv/ci-gha-workflow/pull/155 - cross-build issue

#### Personal updates

* @dmatveev - no major updates, just back from vacation
  * https://github.com/opencv/opencv/pull/23334 - likely needs to be revived soon
  * A DNN-based `infer` - skeleton done, need to complete the implementation, no PR yet
  * https://github.com/opencv/opencv/pull/24379 - OVCV may get attention soon
  * #18 - we've got a GSoC candidate, sadly we don't participate this year (timelines missed for project proposals? https://developers.google.com/open-source/gsoc/timeline) 

@rgarnov - no updates, can follow-up on Anna's PR? https://github.com/opencv/opencv/pull/22469

@TolyaTalamanov - no updates.

### Future agenda

* @TolyaTalamanov - Proposal on loops? Continue discussion.
* @dmatveev - **Concern**: our Fluid kernels may require setting `CPU_BASELINE` manually. Currently the default one is SSE3.1 (?) and we don't do dynamic dispatch in our Fluid kernels. How to make it work out of the box?
